Src/Couchbase/IO/MemoryStreamFactory.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.IO;
+
+namespace Couchbase.IO
+{
+    /// <summary>
+    /// Creates instances of MemoryStreams for use in writing/ reading bytes to/ from the network.  
+    /// </summary>
+    public static class MemoryStreamFactory

```

Comment:
I think it would be good if we could extend this class to be able to setup a custom MemoryStream using configuration (app/web.config) too.

Reply:
Yeah, making this configurable would be a good idea but how would you envision this working? In it's current state, the only way to provide a custom `MemoryStream` is via a delegate. Would the config value accept a static method in the user's code base that was then called via reflection?

An alternative is to expand the scope of this PR and include the Microsoft.IO.RecyclableMemoryStream package and then expose the various configuration knobs to the user (this was the main reason I didn't include the package as part of the PR- finding good default values for the size of the `byte[]` blocks allocated and the maximum size of the pool would be use-case dependent). What do you think?

Reply:
I'd go for putting the MemoryStream type into config then during init create an instance of that type to use in the factory. It's a little bit of reflection but should only happen on start-up. 

We would only get the default values so if a user wanted to tweak specific implementation settings they would have to do it programatically. This way we support any implementation of MemoryStream without locking ourselves into a particular one

Reply:
The issue with providing the name of a type in the config is that object pools like Microsoft.IO.RecyclableMemoryStream require you to create `MemoryStream` instances using a central factory- you can't new up the `MemoryStream` implemented by this package. Or have I misunderstood you, did you mean that the config could accept the name of a method?

Reply:
Ah, the Microsoft.IORecyclableMemoryStream doesn't have a parameterless constructor like System.IO.MemoryStream. That would be tougher to manage via configuration.

I think it's best we leave the configuration piece for now and we can come back to it.

Reply:
Ok that might be for the best. Thanks for the discussion man, most welcome 👍